engine/notifications.js

```diff
@@ -0,0 +1,298 @@
+function register(deviceId, features, endpoint, key, dbNumber) {
+  if (!features || features.length === 0) {
+    return Promise.reject(new Error('No features provided'));
+  }
+
+  return setClient(dbNumber)
+  .then(() => {
+    if (endpoint) {
+      const device = {
+        id: deviceId,
+        endpoint,
+        key: key || '',
+      };
+      redis.hmset(client, `device-${deviceId}`, device);
+      return device;
+    }
+    return redis.hgetall(client, `device-${deviceId}`);
+  })
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('No endpoint provided');
+    }
+  })
+  .then(() => {
+    if (features.indexOf('all') >= 0) {
+      return registerToAll(deviceId, dbNumber);
+    }
+    return Promise.all(
+      features.map(slug => [
+        redis.sadd(client, `${slug}-notifications`, deviceId),
+        redis.sadd(client, `${deviceId}-notifications`, slug)]));
+  })
+  .then(() => sendConfirmationToDevice(deviceId, dbNumber))
+  .then(() => getRegisteredFeatures(deviceId, dbNumber));
+}
+
+// update endpoint for the device
+// required:
+// * machineId
+// * endpoint
+function updateEndpoint(deviceId, endpoint, key, dbNumber) {
+  if (!endpoint) {
+    return Promise.reject(new Error('No endpoint provided'));
+  }
+  return checkDeviceId(deviceId, dbNumber)
+  .then(() => redis.hmset(client, `device-${deviceId}`, 'endpoint', endpoint, 'key', key));
+}
+
+function sendNotifications(feature, payload, isNew, dbNumber) {
+  // make payload a string
+  if (payload && Object.keys(payload).length > 0) {
+    payload = JSON.stringify(payload);
+  }
+  return setClient(dbNumber)
+  .then(() => redis.smembers(client, `${feature}-notifications`))
+  .then(devices => redis.smembers(client, 'all-notifications')
+    .then(all => devices.concat(all)))
+  .then(devices => {
+    if (isNew) {
+      return redis.smembers(client, 'new-notifications')
+      .then(fresh => devices.concat(fresh));
+    }
+    return devices;
+  })
+  .then(devices => Promise.all(
+        devices.map(deviceId => redis.hgetall(client, `device-${deviceId}`))))
+  .then(devices => Promise.all(
+      devices.map(device => {
+        if (device.endpoint.indexOf('https://android.googleapis.com/gcm/send') === 0) {
+          // XXX potential problem if many notifications to the same
+          // machine. An idea to fix it: store an array of messages
+          // instead. Delete after showing it. If a notifications comes
+          // and no payload - just leave it there.
+          return redis.set(client, `${device.id}-payload`, payload)
+          .then(webPush.sendNotification(device.endpoint, ttl));
+        }
+        return webPush.sendNotification(device.endpoint, ttl, device.key, payload);
+      })
+  ));
+}
+
+function getPayload(deviceId, dbNumber) {
+  return setClient(dbNumber)
+  .then(() => redis.get(client, `${deviceId}-payload`))
+  .then(payload => redis.del(client, `${deviceId}-payload`)
+    .then(() => payload));
```

Comment:
Nit: can you move `redis.del` to the following line, so it's clearer what the next `.then` refers to?
```JavaScript
.then(payload =>
  redis.del(client, `${deviceId}-payload`)
  .then(() => payload)
);
```